tests/EnumDynamicReturnTypeExtensionGetValueTest.php

```diff
@@ -0,0 +1,173 @@
+{
+    /** @var EnumDynamicReturnTypeExtension */
+    private $extension;
+
+    private $defaultReturnType = 'array|bool|float|int|string|null';
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->extension = new EnumDynamicReturnTypeExtension();
+
+        // Version < 3.x did not support array values
+        if (method_exists(Enum::class, 'getByName')) {
+            $this->defaultReturnType = 'bool|float|int|string|null';
+        }
+    }
+
+    public function testNullType(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\NullTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', 'null', $this->extension);
+    }
+
+    public function testBoolType(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\BoolTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', 'bool', $this->extension);
+    }
+
+    public function testStringType(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\StrTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', "'str1'|'str2'", $this->extension);
+    }
+
+    public function testIntType(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\IntTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', '0|1', $this->extension);
+    }
+
+    public function testFloatType(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\FloatTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', '1.1|1.2', $this->extension);
+    }
+
+    public function testArrayType(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\ArrayTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', 'array(array())', $this->extension);
+    }
+
+    public function testAllTypes(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\AllTypeEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode(
+            $code,
+            '$enum->getValue()',
+            "1|1.1|'str'|array(null, true, 1, 1.1, 'str', array())|true|null",
+            $this->extension
+        );
+    }
+
+    public function testGeneralizedTypes(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnumPHPStanTest\Assets\BigStrEnum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', 'string', $this->extension);
+    }
+
+    public function testBaseEnum(): void
+    {
+        $code = <<<'CODE'
+<?php
+function f(MabeEnum\Enum $enum) {
+    die;
+}
+CODE;
+
+        $this->processCode($code, '$enum->getValue()', $this->defaultReturnType, $this->extension);
```

Comment:
maybe this should return `never`, because its "not possible" to instantiate abstract class

Reply:
Also it may return never if there is no constant defined in enum. What do you thing?

Reply:
> maybe this should return never, because its "not possible" to instantiate abstract class

You can't instantiate an abstract class but you can type-hint the abstract class which means the `$enum` is an unknown enumerator.

> Also it may return never if there is no constant defined in enum. What do you thing?

You should not be able to instantiate an enumeration without enumerators. So if you have an enumerator this refers to a non empty enumeration in all cases.